docs/tutorials/hpo-workflows/hpo-workflows.rst

```diff
@@ -0,0 +1,334 @@
+
+    .. tab:: RayTorchTrainer
+
+        .. code-block:: python
+
+            class MyRayTorchTrainer(RayTorchTrainer):
+                def __init__(
+                    self,
+                    config: Dict,
+                    strategy: Literal["ddp", "deepspeed"] = "ddp",
+                    name: str | None = None,
+                    logger: Logger | None = None,
+                ) -> None:
+                    super().__init__(config=config, strategy=strategy, name=name, logger=logger)
+
+                def create_model_loss_optimizer(self):
+                    model = resnet18(num_classes=10)
+                    model.conv1 = torch.nn.Conv2d(
+                        1, 64, kernel_size=(7, 7), stride=(1, 1), padding=(3, 3), bias=False
+                    )
+                    # First, define strategy-wise optional configurations
+                    if isinstance(self.strategy, RayDeepSpeedStrategy):
+                        distribute_kwargs = dict(
+                            config_params=dict(
+                                train_micro_batch_size_per_gpu=self.training_config["batch_size"]
+                            )
+                        )
+                    else:
+                        distribute_kwargs = {}
+                    optimizer = Adam(model.parameters(), lr=self.training_config["learning_rate"])
+                    self.model, self.optimizer, _ = self.strategy.distributed(
+                        model, optimizer, **distribute_kwargs
+                    )
+                    self.loss = CrossEntropyLoss()
+
+                def train(self, config, data):
+
+                    ################## This is unique to the RayTorchTrainer #####################
+                    self.training_config = config
+                    self.strategy.init()
+                    self.initialize_logger(
+                        hyperparams=self.training_config, rank=self.strategy.global_rank()
+                    )
+                    self.create_model_loss_optimizer()
+                    self.create_dataloaders(
+                        train_dataset=data[0], validation_dataset=data[1], test_dataset=data[2]
+                    )
+                    ###############################################################################
+
+                    for epoch in range(self.training_config["epochs"]):
+                        if self.strategy.global_world_size() > 1:
+                            self.set_epoch(epoch)
+
+                        train_losses = []
+                        val_losses = []
+
+                        for images, labels in self.train_dataloader:
+                            ############### This is unique to the RayTorchTrainer ##################
+                            if isinstance(self.strategy, RayDeepSpeedStrategy):
+                                device = self.strategy.device()
+                                images, labels = images.to(device), labels.to(device)
```

Comment:
Don't you have to move the images and labels to the correct device even if you aren't using the RayTorchTrainer? It says here that it is unique to the RayTorchTrainer. Do we handle this elsewhere with the regular TorchTrainer?

Reply:
Yeah in the TorchTrainer this is done always, in the RayTorchTrainer only with the RayDeepSpeedStrategy. Do you think it would be clearer to remove this flag/ comment? I could also add a different comment - I liked the idea of the uniform flags, so that it's very easy to see where the changes between the RayTorchTrainer and TorchTrainer are, but maybe that makes it not clear enough what the important changes are?

Reply:
I see what you mean and I think having the flags there is a good idea. I actually missed the `isinstance` here, so it's probably fine the way it is.

Reply:
Okay :)